File: 2024/Leetcode/problem206.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseList(head: ListNode | null): ListNode | null {
    if (!head || !head.next) {
        return head;
    }

    let back = head;
    let curr = head.next;
    let forward;

    if (curr) {
        forward = curr.next;
    } else {
        forward = null;
    }

    back.next = null;

    while (forward) {
        curr.next = back;

        back = curr;
        curr = forward;
        forward = forward.next;
    }

    curr.next = back;

    return curr;
};

File: 2024/Leetcode/problem206.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        back = head
        current = head.next
        front = current.next

        back.next = None

        while front:
            current.next = back

            back = current
            current = front
            front = front.next

        current.next = back
        return current